include/alpaka/atomic/AtomicCpu.hpp

```diff
@@ -0,0 +1,26 @@
+/* Copyright 2021 Andrea Bocci, Felice Pantaleo
+ *
+ * This file is part of alpaka.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#pragma once
+
+#ifndef ALPAKA_DISABLE_ATOMIC_ATOMICREF
+#    include <alpaka/atomic/AtomicAtomicRef.hpp>
+#else
+#    include <alpaka/atomic/AtomicStdLibLock.hpp>
+#endif // ALPAKA_DISABLE_ATOMIC_ATOMICREF
+
+namespace alpaka
+{
+#ifndef ALPAKA_DISABLE_ATOMIC_ATOMICREF
+    using AtomicCpu = AtomicAtomicRef;
+#else
+    using AtomicCpu = AtomicStdLibLock<16>;
+#endif // ALPAKA_DISABLE_ATOMIC_ATOMICREF
```

Comment:
Out of curiosity: What kind of scenarios require defining `ALPAKA_DISABLE_ATOMIC_ATOMICREF`? When is this escape hatch needed? I thought that the `atomic_ref` based implemented could fully replace the lock based one.

Consider also adding this explanation as a comment.

Reply:
As discussed earier in the thread with @jkelling , the [documentation of `boost::atomic_ref`](https://www.boost.org/doc/libs/1_77_0/doc/html/atomic/interface.html#atomic.interface.interface_atomic_ref) says

> The referenced object must have alignment not less than indicated by the `boost::atomic_ref<T>::required_alignment` constant. That constant may be larger than the natural alignment of type `T`. In **Boost.Atomic**, `required_alignment` indicates the alignment at which operations on the object are lock-free; otherwise, if lock-free operations are not possible, `required_alignment` shall not be less than the natural alignment of `T`. 

This is checked at compile time [in each atomic function](https://github.com/alpaka-group/alpaka/pull/1453/files#diff-dccee67d3667c58da43cdba0ebb326a42f05d6fb3b877c3709c6a5491917ebedR39-R42), with
```c++
                static_assert(
                    std::is_trivially_copyable_v<T> && boost::atomic_ref<T>::required_alignment <= alignof(T),
                    "Type not supported by AtomicAtomicRef, please recompile defining "
                    "ALPAKA_DISABLE_ATOMIC_ATOMICREF.");
```
and at runtime by `boost::atomic_ref` itself.

We have not encountered any situation or any POD or non-POD type for which the check failed, but we cannot exclude it could happen on some combination of compiler and architecture.

Reply:
> We have not encountered any situation or any POD or non-POD type for which the check failed

That's the information I was looking for. Given that alpaka atomics are only implemented/tested on the builtin integral and floating point types, I was wondering whether `boost::atomic_ref` fails for anyone of those. Hypothetically, alpaka could support atomics on any kind of type if the user specializes the right templates. So yes, the backend could fail to compile. Whether that is likely or a good idea at all is a different question.